datastore/additional_data/sources/find_that_charity.py

```diff
@@ -0,0 +1,119 @@
+import csv
+import json
+
+import requests
+
+from additional_data.models import OrgInfoCache
+
+
+class FindThatCharitySource(object):
+    """ This is responsible for inputting and outputting data from the
+    FindThatCharity (FTC) organisation info data sources """
+
+    def __init__(self, *args, **kwargs):
+        # A basic internal memory cache to avoid hitting the db on duplicate
+        # requests. Vastly speeds this process up.
+        # OrgInfoCache db typical size is 565,110
+        # This cache object typical size 68,458
+        self._cache = {}
+
+    def update_additional_data(self, grant, additional_data):
+        if "id" not in grant["recipientOrganization"][0]:
+            return
+
+        org_id = grant["recipientOrganization"][0]["id"]
+
+        if "360G-" in org_id:
+            # Not valid org-id
+            return
+
+        # Restart the cache after 300,000 this is approximately
+        # 137.6MiB of memory.
+        if len(self._cache.keys()) > 300000:
+            self._cache = {}
+
+        try:
+            try:
+                # Memory cache because a lot of these are going to be the same
+                org_info = self._cache[org_id]
+                additional_data["recipientOrgInfo"] = org_info
+            except KeyError:
+                org_info = OrgInfoCache.objects.get(org_id=org_id)
```

Comment:
I think `additional_data["recipientOrgInfo"]` should be a list of objects, rather than just one, and we should lookup the org objects based on checking in `org_ids` rather than matching `org_id`. There are instances where a record might need to match based on the non-primary orgid. The main example of this is finding a charity record based on the company number.

I realise this complicates the cache a bit - not sure how you solve this!

Reply:
Yeah, should be easy to change. I don't think the memory cache will be a problem, it is just a key value store. It will be quite a bit slower due to the additional lookups is all.